sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,406 @@
+    """ Bisecting K-Means clustering
+
+    Parameters
+    ----------
+    n_clusters : int, default=8
+        The number of clusters to form as well as the number of
+        centroids to generate.
+
+     n_init : int, default=10
+        Number of time the k-means algorithm will be run with different
+        centroid seeds in each bisection.
+        The final results will be the best output of n_init consecutive
+        runs in terms of inertia.
+
+    random_state : int, RandomState instance or None, default=None
+        Determines random number generation for centroid initialization. Use
+        an int to make the randomness deterministic.
+        See :term:`Glossary <random_state>`.
+
+    max_iter : int, default=300
+        Maximum number of iterations of the k-means algorithm for a
+        single run.
+
+    Verbose : int, default=0
+        Verbosity mode.
+
+    tol : float, default=1e-4
+        Relative tolerance with regards to Frobenius norm of the difference
+        in the cluster centers of two consecutive iterations to declare
+        convergence.
+
+    copy_x : bool, default=True
+        When pre-computing distances it is more numerically accurate to center
+        the data first. If copy_x is True (default), then the original data is
+        not modified. If False, the original data is modified, and put back
+        before the function returns, but small numerical differences may be
+        introduced by subtracting and then adding the data mean. Note that if
+        the original data is not C-contiguous, a copy will be made even if
+        copy_x is False. If the original data is sparse, but not in CSR format,
+        a copy will be made even if copy_x is False.
+
+    algorithm : {"auto", "full", "elkan"}, default="auto"
+        K-means algorithm to use. The classical EM-style algorithm is "full".
+        The "elkan" variation is more efficient on data with well-defined
+        clusters, by using the triangle inequality. However it's more memory
+        intensive due to the allocation of an extra array of shape
+        (n_samples, n_clusters).
+        For now "auto" (kept for backward compatibiliy) chooses "elkan" but it
+        might change in the future for a better heuristic.
+
+    Attributes
+    ----------
+    cluster_centers_ : ndarray of shape (n_clusters, n_features)
+        Coordinates of cluster centers. If the algorithm stops before fully
+        converging (see ``tol`` and ``max_iter``), these will not be
+        consistent with ``labels_``.
+    labels_ : ndarray of shape (n_samples,)
+        Labels of each point
+    inertia_ : float
+        Sum of squared distances of samples to their closest cluster center,
+        weighted by the sample weights if provided.
+    n_iter_ : int
+        Number of iterations run.
+
+    Notes
+    -----
+    That algorithm will not work if n_cluster is smaller than 2.
+
+    Also it might be inefficient when n_cluster is equal to 2
```

Comment:
If there is a good reference for this algorithm, here might be a good place to add it.

Reply:
FYI @GaelVaroquaux  we're moving towards having the references in the user guide (this PR will need a user guide entry BTW) instead of the docstrings: when there are 5 references in a docstring it's impossible to know which one referes to which part of the algorithm. When the references are in the UG, we can clearly link each section to its relevant reference / paper. See e.g. how it's done for the SGD user guide: https://scikit-learn.org/stable/modules/sgd.html

Reply:
Well that's a pity. It's really useful to look at a docstring and know exactly which algorithm it implements (there are so many grey areas).

My comment was actually driven by those grey areas. I would genuinely like to know specifically which algorithm the code is implementing as I am reading the code.

Reply:
Docstrings often have 5+ references with zero context. Some of these references are about the algoirthm in general, and some can be about a single detail / parameter / solver. Unless you already know what you're looking for, finding which paper you should read is hard. Having the refs in the UG with proper links helps: https://github.com/scikit-learn/scikit-learn/pull/16296